# Session 11: Advanced Retrieval with LangChain

## Learning Objectives:

- Understand and implement multiple retrieval strategies for RAG
- Compare naive, BM25, multi-query, parent-document, contextual compression, ensemble, and semantic chunking approaches
- Build RAG chains over a health and wellness knowledge base using LangChain and QDrant

In the following notebook, we'll explore various methods of advanced retrieval using LangChain!

We'll touch on:

- Naive Retrieval
- Best-Matching 25 (BM25)
- Multi-Query Retrieval
- Parent-Document Retrieval
- Contextual Compression (a.k.a. Rerank)
- Ensemble Retrieval
- Semantic chunking

We'll also discuss how these methods impact performance on our set of documents with a simple RAG chain.

There will be two breakout rooms:

- 🤝 Breakout Room Part #1
  - Task 1: Getting Dependencies!
  - Task 2: Data Collection and Preparation
  - Task 3: Setting Up QDrant!
  - Task 4-10: Retrieval Strategies
- 🤝 Breakout Room Part #2
  - Activity: Evaluate with Ragas

---

# 🤝 Breakout Room Part #1

## Task 1: Getting Dependencies!

We're going to need a few specific LangChain community packages, like OpenAI (for our [LLM](https://platform.openai.com/docs/models) and [Embedding Model](https://platform.openai.com/docs/guides/embeddings)) and Cohere (for our [Reranker](https://cohere.com/rerank)).

We'll also provide our OpenAI key, as well as our Cohere API key.

> NOTE: Create a `.env` file in this directory with `OPENAI_API_KEY` and `COHERE_API_KEY` to avoid being prompted each time.

In [1]:
import os
import getpass
from dotenv import load_dotenv
from langsmith import EvaluationResult

load_dotenv()

if not os.environ.get("OPENAI_API_KEY"):
    os.environ["OPENAI_API_KEY"] = getpass.getpass("Enter your OpenAI API Key:")

In [2]:
if not os.environ.get("COHERE_API_KEY"):
    os.environ["COHERE_API_KEY"] = getpass.getpass("Cohere API Key:")

## Task 2: Data Collection and Preparation

We'll be using our Health and Wellness Guide - a comprehensive resource covering exercise, nutrition, sleep, stress management, habits, and common health concerns.

### Data Preparation

We'll load the wellness guide as a single document, then split it into smaller chunks using a `RecursiveCharacterTextSplitter` for our vector store. We also keep the raw (unsplit) document for use with the Parent Document Retriever and Semantic Chunker later.

In [3]:
from langchain_community.document_loaders import TextLoader
from langchain_text_splitters import RecursiveCharacterTextSplitter

loader = TextLoader("data/HealthWellnessGuide.txt")
raw_docs = loader.load()

text_splitter = RecursiveCharacterTextSplitter(chunk_size=500, chunk_overlap=50)
wellness_docs = text_splitter.split_documents(raw_docs)

Let's verify our data was loaded and split correctly!

In [4]:
print(f"Raw documents: {len(raw_docs)}")
print(f"Split chunks: {len(wellness_docs)}")
print(f"\nExample chunk:\n{wellness_docs[0]}")

Raw documents: 1
Split chunks: 45

Example chunk:
page_content='The Personal Wellness Guide
A Comprehensive Resource for Health and Well-being

PART 1: EXERCISE AND MOVEMENT

Chapter 1: Understanding Exercise Basics

Exercise is one of the most important things you can do for your health. Regular physical activity can improve your brain health, help manage weight, reduce the risk of disease, strengthen bones and muscles, and improve your ability to do everyday activities.' metadata={'source': 'data/HealthWellnessGuide.txt'}


## Task 3: Setting up QDrant!

Now that we have our documents, let's create a QDrant VectorStore with the collection name "wellness_guide".

We'll leverage OpenAI's [`text-embedding-3-small`](https://openai.com/blog/new-embedding-models-and-api-updates) because it's a very powerful (and low-cost) embedding model.

> NOTE: We'll be creating additional vectorstores where necessary, but this pattern is still extremely useful.

In [5]:
from langchain_qdrant import QdrantVectorStore
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

vectorstore = QdrantVectorStore.from_documents(
    wellness_docs,
    embeddings,
    location=":memory:",
    collection_name="wellness_guide",
)

## Task 4: Naive RAG Chain

Since we're focusing on the "R" in RAG today - we'll create our Retriever first.

### R - Retrieval

This naive retriever will simply look at each review as a document, and use cosine-similarity to fetch the 10 most relevant documents.

> NOTE: We're choosing `10` as our `k` here to provide enough documents for our reranking process later

In [6]:
naive_retriever = vectorstore.as_retriever(search_kwargs={"k" : 10})

### A - Augmented

We're going to go with a standard prompt for our simple RAG chain today! Nothing fancy here, we want this to mostly be about the Retrieval process.

In [7]:
from langchain_core.prompts import ChatPromptTemplate

RAG_TEMPLATE = """\
You are a helpful and kind assistant. Use the context provided below to answer the question.

If you do not know the answer, or are unsure, say you don't know.

Query:
{question}

Context:
{context}
"""

rag_prompt = ChatPromptTemplate.from_template(RAG_TEMPLATE)

### G - Generation

We're going to leverage `gpt-4.1-nano` as our LLM today, as - again - we want this to largely be about the Retrieval process.

In [8]:
from langchain_openai import ChatOpenAI

chat_model = ChatOpenAI(model="gpt-4.1-nano")

### LCEL RAG Chain

We're going to use LCEL to construct our chain.

> NOTE: This chain will be exactly the same across the various examples with the exception of our Retriever!

In [9]:
from langchain_core.runnables import RunnablePassthrough
from operator import itemgetter
from langchain_core.output_parsers import StrOutputParser

naive_retrieval_chain = (
    # INVOKE CHAIN WITH: {"question" : "<<SOME USER QUESTION>>"}
    # "question" : populated by getting the value of the "question" key
    # "context"  : populated by getting the value of the "question" key and chaining it into the base_retriever
    {"context": itemgetter("question") | naive_retriever, "question": itemgetter("question")}
    # "context"  : is assigned to a RunnablePassthrough object (will not be called or considered in the next step)
    #              by getting the value of the "context" key from the previous step
    | RunnablePassthrough.assign(context=itemgetter("context"))
    # "response" : the "context" and "question" values are used to format our prompt object and then piped
    #              into the LLM and stored in a key called "response"
    # "context"  : populated by getting the value of the "context" key from the previous step
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

Let's see how this simple chain does on a few different prompts.

> NOTE: You might think that we've cherry picked prompts that showcase the individual skill of each of the retrieval strategies - you'd be correct!

In [10]:
naive_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'Exercises that can help with lower back pain include:\n\n- **Cat-Cow Stretch:** Start on hands and knees, alternate between arching your back up (cat) and letting it sag down (cow). Do 10-15 repetitions.\n\n- **Bird Dog:** From hands and knees, extend opposite arm and leg while keeping your core engaged. Hold for 5 seconds, then switch sides. Do 10 repetitions per side.\n\n- **Pelvic Tilts:** Lie on your back with knees bent, flatten your back against the floor by tightening your abs and tilting the pelvis up slightly. Hold for 10 seconds and repeat 8-12 times.\n\n- **Partial Crunches:** Lie on your back with knees bent, cross arms over chest, tighten stomach muscles, and raise shoulders off the floor. Hold briefly, then lower. Do 8-12 repetitions.\n\n- **Knee-to-Chest Stretch:** Lie on your back, pull one knee toward your chest while keeping the other foot flat. Hold for 15-30 seconds, then switch legs.\n\nThese exercises, performed gently and regularly, can help alleviate lower back

In [11]:
naive_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

'Sleep has a significant impact on overall health. Adequate sleep (7-9 hours per night) is essential for physical health, mental well-being, and cognitive function. During sleep, the body repairs tissues, consolidates memories, and releases hormones that help regulate growth and appetite. Good sleep supports a healthy immune system, enhances mood, improves concentration, and contributes to better recovery from illnesses. Conversely, poor sleep or sleep disorders like insomnia can negatively affect these areas, highlighting the importance of maintaining good sleep hygiene and creating an optimal sleep environment for overall wellness.'

In [12]:
naive_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress and headaches include:\n\n- Drinking water and staying hydrated\n- Applying cold or warm compresses to the head or neck\n- Resting in a dark, quiet room\n- Gentle massage of the temples and neck\n- Using essential oils such as peppermint or lavender\n- Taking small amounts of caffeine (with caution)\n- Practicing relaxation techniques like deep breathing, progressive muscle relaxation, or grounding exercises\n- Taking short walks, especially in nature\n- Listening to calming music\n\nThese methods can help alleviate stress and manage headache symptoms naturally.'

Overall, this is not bad! Let's see if we can make it better!

## Task 5: Best-Matching 25 (BM25) Retriever

Taking a step back in time - [BM25](https://www.nowpublishers.com/article/Details/INR-019) is based on [Bag-Of-Words](https://en.wikipedia.org/wiki/Bag-of-words_model) which is a sparse representation of text.

In essence, it's a way to compare how similar two pieces of text are based on the words they both contain.

This retriever is very straightforward to set-up! Let's see it happen down below!


In [13]:
from langchain_community.retrievers import BM25Retriever

bm25_retriever = BM25Retriever.from_documents(wellness_docs)

We'll construct the same chain - only changing the retriever.

In [14]:
bm25_retrieval_chain = (
    {"context": itemgetter("question") | bm25_retriever, "question": itemgetter("question")}
    | RunnablePassthrough.assign(context=itemgetter("context"))
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

Let's look at the responses!

In [15]:
bm25_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'Exercises that can help with lower back pain include:\n\n- Cat-Cow Stretch: Start on your hands and knees, then alternate between arching your back up (cat) and letting it sag down (cow). Perform 10-15 repetitions.\n- Bird Dog: From hands and knees, extend opposite arm and leg while keeping your core engaged. Hold each extension for 5 seconds, then switch sides. Do 10 repetitions per side.\n- Pelvic Tilts: Lie on your back with knees bent, flatten your lower back against the floor by tightening your abdominal muscles and tilting your pelvis slightly upward. Hold for 10 seconds and repeat 8-12 times.\n\nThese exercises are gentle stretching and strengthening movements that can help alleviate and prevent lower back pain.'

In [16]:
bm25_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

'Sleep plays a vital role in overall health. Practicing good sleep hygiene, such as maintaining a consistent sleep schedule, creating a comfortable sleep environment (cool, dark, quiet), and establishing relaxing routines before bed, helps improve sleep quality. Adequate sleep supports immune function, mental health, and vital bodily processes. Conversely, poor sleep or insomnia can negatively impact overall wellness, emphasizing the importance of healthy sleep habits for maintaining good health.'

In [17]:
bm25_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress and headaches include relaxation techniques such as meditation and deep breathing exercises, progressive muscle relaxation, and herbal teas like chamomile or valerian root. Additionally, addressing triggers such as dehydration, poor sleep, or eye strain can help prevent headaches. For stress, practicing mindfulness and incorporating relaxation methods into your routine can be beneficial.'

|It's not clear that this is better or worse, if only we had a way to test this (SPOILERS: We do, the second half of the notebook will cover this)

### ❓ Question #1:

Give an example query where BM25 is better than embeddings and justify your answer.

##### Answer:

- BM25 will be better than embeddings (vector search) in scenarios where the input query contains rare words or text that may not be embedded well into any embedding model
- IE: say you are a nurse asking a question particular to an ICD-10 code (codes that identify diseases) in your query to the RAG prompt
  - since BM25 internally uses `TF-IDF`, it will actually give a high score to chunks of data that contain this ICD-10 code and score it highly because ICD-10 codes are pretty rare
  - in an embedding model, the ICD-10 code may not even exist since they are usually very rare (you may need to utilize an embedding model that is specifically trained/used on healthcare/electronic medical record data)
  - so anytime a rare keyword is important to an input query or question (any sort of keyword search), BM25 will do much better than an embedding model because the semantic meaning of the query doesn't matter as much as the keywords do
  - BM25 will always except at Control-F type of queries

## Task 6: Contextual Compression (Using Reranking)

Contextual Compression is a fairly straightforward idea: We want to "compress" our retrieved context into just the most useful bits.

There are a few ways we can achieve this - but we're going to look at a specific example called reranking.

The basic idea here is this:

- We retrieve lots of documents that are very likely related to our query vector
- We "compress" those documents into a smaller set of *more* related documents using a reranking algorithm.

We'll be leveraging Cohere's Rerank model for our reranker today!

All we need to do is the following:

- Create a basic retriever
- Create a compressor (reranker, in this case)

That's it!

Let's see it in the code below!

In [18]:
from langchain.retrievers.contextual_compression import ContextualCompressionRetriever
from langchain_cohere import CohereRerank

compressor = CohereRerank(model="rerank-v3.5")
compression_retriever = ContextualCompressionRetriever(
    base_compressor=compressor, base_retriever=naive_retriever
)

Let's create our chain again, and see how this does!

In [19]:
contextual_compression_retrieval_chain = (
    {"context": itemgetter("question") | compression_retriever, "question": itemgetter("question")}
    | RunnablePassthrough.assign(context=itemgetter("context"))
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

In [20]:
contextual_compression_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'Based on the provided information, exercises that can help with lower back pain include:\n\n1. **Cat-Cow Stretch**: Start on your hands and knees. Alternate between arching your back up (like a cat) and letting it sag down (like a cow). Perform 10-15 repetitions.\n\n2. **Bird Dog**: On your hands and knees, extend opposite arm and leg while keeping your core engaged. Hold each extension for about 5 seconds, then switch sides. Do 10 repetitions per side.\n\n3. **Pelvic Tilts**: Lie on your back with knees bent. Flatten your back against the floor by tightening your abdominal muscles and tilting your pelvis slightly upward. Hold for 10 seconds and repeat 8-12 times.\n\nThese gentle stretching and strengthening exercises can help alleviate lower back discomfort and may also prevent future episodes.'

In [21]:
contextual_compression_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

"Sleep has a significant impact on overall health. It is essential for physical health, mental well-being, and cognitive function. During sleep, the body repairs tissues, consolidates memories, and releases hormones that regulate growth and appetite. Adequate sleep, typically 7-9 hours per night, supports the body's healing processes, helps maintain a healthy immune system, and enhances mental clarity and emotional stability. Conversely, insufficient or poor-quality sleep can impair these functions and negatively affect overall health."

In [22]:
contextual_compression_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress and headaches include drinking water to stay hydrated, applying cold or warm compresses to the head or neck, resting in a dark and quiet room, giving gentle massage to the temples and neck, using essential oils like peppermint or lavender, maintaining a regular sleep schedule, practicing deep breathing, engaging in progressive muscle relaxation, using grounding techniques, taking short walks in nature, and listening to calming music.'

We'll need to rely on something like Ragas to help us get a better sense of how this is performing overall - but it "feels" better!

## Task 7: Multi-Query Retriever

Typically in RAG we have a single query - the one provided by the user.

What if we had....more than one query!

In essence, a Multi-Query Retriever works by:

1. Taking the original user query and creating `n` number of new user queries using an LLM.
2. Retrieving documents for each query.
3. Using all unique retrieved documents as context

So, how is it to set-up? Not bad! Let's see it down below!



In [23]:
from langchain.retrievers.multi_query import MultiQueryRetriever

multi_query_retriever = MultiQueryRetriever.from_llm(
    retriever=naive_retriever, llm=chat_model
) 

In [24]:
multi_query_retrieval_chain = (
    {"context": itemgetter("question") | multi_query_retriever, "question": itemgetter("question")}
    | RunnablePassthrough.assign(context=itemgetter("context"))
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

In [25]:
multi_query_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'Exercises that can help with lower back pain include:\n\n- **Cat-Cow Stretch:** Start on your hands and knees, then alternate between arching your back upward (cat) and letting it sag down (cow). Perform 10-15 repetitions.\n- **Bird Dog:** From hands and knees, extend opposite arm and leg simultaneously, keeping your core engaged. Hold for 5 seconds, then switch sides. Do 10 repetitions per side.\n- **Partial Crunches:** Lie on your back with knees bent, arms crossed over your chest, tighten your stomach muscles, and lift your shoulders off the floor. Hold briefly and then lower. Do 8-12 repetitions.\n- **Knee-to-Chest Stretch:** Lie on your back, pull one knee toward your chest while keeping the other foot flat. Hold for 15-30 seconds, then switch legs.\n- **Pelvic Tilts:** Lie on your back with knees bent, flatten your back against the floor by tightening your abs and tilting your pelvis upward. Hold for 10 seconds and repeat 8-12 times.\n\nThese exercises, when done properly, can h

In [26]:
multi_query_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

'Sleep has a significant impact on overall health. According to the information provided, sleep is crucial for physical health, mental well-being, and cognitive function. During sleep, your body repairs tissues, consolidates memories, and releases hormones that regulate growth and appetite. Adequate sleep—typically 7-9 hours for adults—is essential for maintaining a strong immune system, supporting mental health, and ensuring optimal cognitive performance. Poor sleep or sleep problems like insomnia can negatively affect health, leading to issues such as stress, fatigue, and decreased immunity. Therefore, establishing good sleep hygiene and creating a sleep-friendly environment are vital practices to promote overall health.'

In [27]:
multi_query_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress and headaches include:\n\n- Deep breathing exercises (e.g., inhale for 4 counts, hold for 4, exhale for 4)\n- Progressive muscle relaxation (tensing and releasing muscle groups)\n- Grounding techniques (naming things you see, hear, feel, smell, and taste)\n- Taking short walks, preferably in nature\n- Listening to calming music\n- Drinking water to stay hydrated\n- Applying cold or warm compresses to the head or neck\n- Resting in a dark, quiet room\n- Gentle massage of temples and neck\n- Using essential oils like peppermint or lavender\n- Maintaining a regular sleep schedule\n\nThese approaches can help alleviate stress and headaches naturally without medication.'

### ❓ Question #2:

Explain how generating multiple reformulations of a user query can improve recall.

##### Answer:
- multiple reformulations of a user query will generate slightly different questions with different wording or even different terms/vocabulary
  - the different wording will cast a larger net of documents to retrieve (documents that could have been missed by original query) which could help increase recall
- since each query has its own set of documents retrieved, there's a high probability that the union of documents will be larger than what the original query would have retrieved; and as long as these documents are semantically relevant to the original query, it should increase the recall
- each reformulated query emphasizes different aspects of the original query which could capture new documents that also emphasize those specific angles (that the original query would not have captured)
- generating reformulations of the same question is essentially compensating for the context "compression" that naturally occurs with embeddings; the reformulations allow us to search in other semantic neighborhoods that may also have relevant documents to our original query

## Task 8: Parent Document Retriever

A "small-to-big" strategy - the Parent Document Retriever works based on a simple strategy:

1. We split the full document into large "parent" chunks (e.g. 2000 characters).
2. Each parent chunk is further split into smaller "child" chunks (e.g. 400 characters).
3. The child chunks are stored in a VectorStore, while the parent chunks are stored in an in-memory docstore.
4. When we query our Retriever, we do a similarity search comparing our query vector to the child chunks.
5. Instead of returning the child chunks, we return their associated parent chunks.

The basic idea is:

- **Search** for small, focused chunks (better semantic matching)
- **Return** big chunks (richer surrounding context)

The intuition is that we're likely to find the most relevant information by limiting the amount of semantic information encoded in each embedding vector - but we're likely to miss relevant surrounding context if we only use that information.

Let's start by defining our parent and child splitters.

In [28]:
from langchain.retrievers import ParentDocumentRetriever
from langchain.storage import InMemoryStore
from langchain_text_splitters import RecursiveCharacterTextSplitter
from qdrant_client import QdrantClient, models

parent_splitter = RecursiveCharacterTextSplitter(chunk_size=2000, chunk_overlap=200)
child_splitter = RecursiveCharacterTextSplitter(chunk_size=400, chunk_overlap=50)

We'll need to set up a new QDrant vectorstore - and we'll use another useful pattern to do so!

> NOTE: We are manually defining our embedding dimension, you'll need to change this if you're using a different embedding model.

In [29]:
from langchain_qdrant import QdrantVectorStore

client = QdrantClient(location=":memory:")

client.create_collection(
    collection_name="wellness_parent_child",
    vectors_config=models.VectorParams(size=1536, distance=models.Distance.COSINE)
)

parent_document_vectorstore = QdrantVectorStore(
    collection_name="wellness_parent_child", embedding=OpenAIEmbeddings(model="text-embedding-3-small"), client=client
)

Now we can create our `InMemoryStore` that will hold our "parent documents" - and build our retriever!

In [30]:
store = InMemoryStore()

parent_document_retriever = ParentDocumentRetriever(
    vectorstore=parent_document_vectorstore,
    docstore=store,
    child_splitter=child_splitter,
    parent_splitter=parent_splitter,
)

By default, this is empty as we haven't added any documents - let's add some now!

In [31]:
parent_document_retriever.add_documents(raw_docs, ids=None)

We'll create the same chain we did before - but substitute our new `parent_document_retriever`.

In [32]:
parent_document_retrieval_chain = (
    {"context": itemgetter("question") | parent_document_retriever, "question": itemgetter("question")}
    | RunnablePassthrough.assign(context=itemgetter("context"))
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

Let's give it a whirl!

In [33]:
parent_document_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'To help alleviate lower back pain, the following exercises are recommended:\n\n- **Cat-Cow Stretch:** Start on hands and knees; arch your back up (cat) and then let it sag down (cow). Repeat 10-15 times.\n- **Bird Dog:** From hands and knees, extend opposite arm and leg while engaging your core. Hold for 5 seconds, then switch sides. Do 10 repetitions per side.\n- **Partial Crunches:** Lie on your back with knees bent, cross arms over your chest, tighten your stomach muscles, and lift your shoulders off the floor. Repeat 8-12 times.\n- **Knee-to-Chest Stretch:** Lie on your back and pull one knee toward your chest, keeping the other foot flat on the floor. Hold for 15-30 seconds, then switch legs.\n- **Pelvic Tilts:** Lie on your back with knees bent, flatten your back against the floor by tightening your abs and tilting your pelvis slightly upward. Hold for 10 seconds and repeat 8-12 times.\n\nThese exercises, when performed gently and regularly, can help reduce lower back pain and p

In [35]:
parent_document_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

'Sleep has a significant impact on overall health. It is crucial for physical health, mental well-being, and cognitive function. During sleep, the body repairs tissues, consolidates memories, and releases hormones that help regulate growth and appetite. Adults typically need 7-9 hours of quality sleep per night in order for these processes to occur effectively. Good sleep hygiene—such as maintaining a consistent sleep schedule, creating a relaxing bedtime routine, and optimizing the sleep environment—can improve sleep quality. Adequate sleep supports vital bodily functions, enhances mood, boosts immune function, and promotes better mental clarity and learning. Conversely, poor sleep or sleep disturbances can lead to various health issues, including increased risk of chronic conditions, impaired immune response, and mental health challenges.'

In [36]:
parent_document_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress and headaches include:\n\n- Deep breathing exercises to promote relaxation\n- Progressive muscle relaxation to reduce physical tension\n- Grounding techniques like identifying things you see, hear, feel, smell, and taste to center yourself\n- Taking short walks, preferably in nature, to clear the mind\n- Listening to calming music\n- Using essential oils such as peppermint or lavender, which can help alleviate headache symptoms\n- Ensuring adequate hydration by drinking plenty of water\n- Resting in a dark, quiet room to reduce headache symptoms\n- Gentle massage of the temples and neck to relieve tension\n- Maintaining a regular sleep schedule and practicing good sleep hygiene\n\nThese approaches focus on calming the mind and body naturally, which can help manage stress and headaches effectively.'

Overall, the performance *seems* largely the same. We can leverage a tool like [Ragas]() to more effectively answer the question about the performance.

## Task 9: Ensemble Retriever

In brief, an Ensemble Retriever simply takes 2, or more, retrievers and combines their retrieved documents based on a rank-fusion algorithm.

In this case - we're using the [Reciprocal Rank Fusion](https://plg.uwaterloo.ca/~gvcormac/cormacksigir09-rrf.pdf) algorithm.

Setting it up is as easy as providing a list of our desired retrievers - and the weights for each retriever.

In [37]:
from langchain.retrievers import EnsembleRetriever

retriever_list = [bm25_retriever, naive_retriever, parent_document_retriever, compression_retriever, multi_query_retriever]
equal_weighting = [1/len(retriever_list)] * len(retriever_list)

ensemble_retriever = EnsembleRetriever(
    retrievers=retriever_list, weights=equal_weighting
)

We'll pack *all* of these retrievers together in an ensemble.

In [38]:
ensemble_retrieval_chain = (
    {"context": itemgetter("question") | ensemble_retriever, "question": itemgetter("question")}
    | RunnablePassthrough.assign(context=itemgetter("context"))
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

Let's look at our results!

In [39]:
ensemble_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'Exercises that can help with lower back pain include:\n\n- **Cat-Cow Stretch:** Start on your hands and knees, alternate between arching your back up (like a cat) and letting it sag down (like a cow). Perform 10-15 repetitions.\n\n- **Bird Dog:** From hands and knees, extend opposite arm and leg while keeping your core engaged. Hold for 5 seconds, then switch sides. Do 10 repetitions per side.\n\n- **Pelvic Tilts:** Lie on your back with knees bent, tighten your abs, and tilt your pelvis slightly upward to flatten your back against the floor. Hold for 10 seconds and repeat 8-12 times.\n\n- **Partial Crunches:** Lie on your back with knees bent, cross your arms over your chest, and tighten your stomach muscles to lift your shoulders off the floor. Hold briefly, then lower down. Do 8-12 repetitions.\n\n- **Knee-to-Chest Stretch:** Lie on your back, pull one knee toward your chest while keeping the other foot flat on the floor. Hold for 15-30 seconds, then switch legs.\n\nThese exercises

In [40]:
ensemble_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

'Sleep plays a vital role in maintaining overall health. It supports physical well-being by allowing the body to repair tissues and regenerate, especially during deep sleep stages. Sleep also enhances mental health by helping consolidate memories and regulate emotions. Additionally, sufficient sleep influences cognitive function, mood, hormonal balance (such as growth and appetite hormones), and immune function. Poor or inadequate sleep can lead to various health problems, including increased stress, weakened immunity, cognitive impairments, and higher risk of chronic conditions. Therefore, maintaining good sleep hygiene and creating an optimal sleep environment are important for overall health and wellness.'

In [41]:
ensemble_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress and headaches include:\n\n- Deep breathing exercises\n- Progressive muscle relaxation\n- Grounding techniques (naming sights, sounds, feelings, smells, tastes)\n- Taking short walks, preferably in nature\n- Listening to calming music\n- Applying cold or warm compresses to the head or neck\n- Resting in a dark, quiet room\n- Gentle massage of temples and neck\n- Using peppermint or lavender essential oils\n- Maintaining a regular sleep schedule\n- Staying well-hydrated by drinking water\n- Managing stress through regular exercise, social support, hobbies, and mindfulness practices\n\nThese approaches can help relieve stress and headaches naturally and promote overall well-being.'

## Task 10: Semantic Chunking

While this is not a retrieval method - it *is* an effective way of increasing retrieval performance on corpora that have clean semantic breaks in them.

Essentially, Semantic Chunking is implemented by:

1. Embedding all sentences in the corpus.
2. Combining or splitting sequences of sentences based on their semantic similarity based on a number of [possible thresholding methods](https://python.langchain.com/docs/how_to/semantic-chunker/):
  - `percentile`
  - `standard_deviation`
  - `interquartile`
  - `gradient`
3. Each sequence of related sentences is kept as a document!

Let's see how to implement this!

We'll use the `percentile` thresholding method for this example which will:

Calculate all distances between sentences, and then break apart sequences of setences that exceed a given percentile among all distances.

In [42]:
from langchain_experimental.text_splitter import SemanticChunker

semantic_chunker = SemanticChunker(
    embeddings,
    breakpoint_threshold_type="percentile"
)

Now we can split our documents.

In [43]:
semantic_documents = semantic_chunker.split_documents(raw_docs)

Let's create a new vector store.

In [44]:
semantic_vectorstore = QdrantVectorStore.from_documents(
    semantic_documents,
    embeddings,
    location=":memory:",
    collection_name="wellness_guide_semantic_chunks"
)

We'll use naive retrieval for this example.

In [45]:
semantic_retriever = semantic_vectorstore.as_retriever(search_kwargs={"k" : 10})

Finally we can create our classic chain!

In [46]:
semantic_retrieval_chain = (
    {"context": itemgetter("question") | semantic_retriever, "question": itemgetter("question")}
    | RunnablePassthrough.assign(context=itemgetter("context"))
    | {"response": rag_prompt | chat_model, "context": itemgetter("context")}
)

And view the results!

In [47]:
semantic_retrieval_chain.invoke({"question" : "What exercises can help with lower back pain?"})["response"].content

'Exercises that can help with lower back pain include:\n\n- **Cat-Cow Stretch:** Start on hands and knees, alternate between arching your back up (cat) and sagging it down (cow). Do 10-15 repetitions.\n- **Partial Crunches:** Lie on your back with knees bent, cross arms over chest, tighten stomach muscles, and lift shoulders off the floor. Perform 8-12 repetitions.\n- **Knee-to-Chest Stretch:** Lie on your back, pull one knee toward your chest, hold for 15-30 seconds, then switch legs.\n- **Pelvic Tilts:** Lie on your back with knees bent, flatten your back against the floor by tightening your abs and tilting your pelvis slightly. Hold for 10 seconds, repeat 8-12 times.\n\nThese gentle stretching and strengthening exercises are recommended to alleviate lower back discomfort and prevent future episodes.'

In [48]:
semantic_retrieval_chain.invoke({"question" : "How does sleep affect overall health?"})["response"].content

'Sleep plays a vital role in overall health by supporting physical, mental, and cognitive functions. During sleep, the body repairs tissues, regenerates cells, consolidates memories, and releases important hormones that regulate growth and appetite. Adults generally need 7-9 hours of sleep per night, with sleep occurring in cycles that include REM and non-REM stages, each contributing to different restorative processes. Good sleep quality, maintained through proper sleep hygiene practices like a consistent routine, a comfortable environment, and limiting screens before bed, is crucial for maintaining optimal health. Poor sleep can lead to symptoms such as fatigue, headaches, and low energy, and over time, can increase the risk of chronic conditions like obesity, diabetes, cardiovascular disease, and mental health disorders.'

In [49]:
semantic_retrieval_chain.invoke({"question" : "What are some natural remedies for stress and headaches?"})["response"].content

'Some natural remedies for stress include practicing deep breathing exercises, progressive muscle relaxation, grounding techniques (such as naming things you see, hear, feel, smell, and taste), taking short walks in nature, and listening to calming music. For headaches, natural remedies include staying hydrated by drinking water, applying cold or warm compresses to the head or neck, resting in a dark, quiet room, gentle massage of the temples and neck, using essential oils like peppermint or lavender, and maintaining a regular sleep schedule.'

### ❓ Question #3:

If sentences are short and highly repetitive (e.g., FAQs), how might semantic chunking behave, and how would you adjust the algorithm?

##### Answer:
- the problem with short and highly repetitive sentences (like FAQs) is that many consecutive sentences might be very similar to each other (especially in the embedding space) which will make it difficult to find chunk boundaries with semantic chunking
  - the distances from similar sentences will be very small, so thresholding types such as percentile, standard deviation, and interquartile will all have issues finding chunk boundaries (there will be too few boundaries)
    - percentile considers the all the distances between consecutive sentences and creates a chunk boundary past a certain percentile threshold (IE 95th percentile)
      - but with FAQs, all distances will be very similar (close to uniformly distributed) so the 95th percentile might still be very low causing not enough splits
    - the standard deviation threshold type calculates the mean and standard deviation of all consecutive sentence distances and creates a chunk boundary when a distance exceeds `mean + k * std dev` (the parameter k controls the sensitivity)
      - this is more adaptative to your document but with FAQs, the standard deviation will be low so small variations could lead to over-splitting
    - the interquartile threshold type utilizes the middle 50th percentile of distances (`IQR = Q3 - Q1`) and creates a chunk boundary when distances are greater than `Q3 + k * IQR`
      - this is similar to the issue with standard deviaton; FAQs will have a very small IQR due to so many similar chunks so the IQR method may be too sensitive and create too many chunk boundaries
- I would adjust the algorithm by utilizing the gradient threshold type
  - instead of looking at absolute distances, the gradient method looks at the rate of change in distances
  - it creates a chunk boundary when there is a sudden change (or rate of change) in distances (it detects jumps in semantic dissimilarity)
  - gradient thresholding will perform better than the other threshold types for finding meaningful chunk boundaries and topic changes with repetitive sentences (like FAQs) since it is looking at the rate of change in distance instead of absolute distance

---

# 🤝 Breakout Room Part #2

### 🏗️ Activity #1:

Your task is to evaluate the various Retriever methods against each other.

You are expected to:

1. Create a "golden dataset"
 - Use Synthetic Data Generation (powered by Ragas, or otherwise) to create this dataset
2. Evaluate each retriever with *retriever specific* Ragas metrics
 - Semantic Chunking is not considered a retriever method and will not be required for marks, but you may find it useful to do a "semantic chunking on" vs. "semantic chunking off" comparison between them
3. Compile these in a list and write a small paragraph about which is best for this particular data and why.

Your analysis should factor in:
  - Cost
  - Latency
  - Performance

> NOTE: This is **NOT** required to be completed in class. Please spend time in your breakout rooms creating a plan before moving on to writing code.

##### HINTS:

- LangSmith provides detailed information about latency and cost.

In [1]:
### YOUR CODE HERE
from act1 import run

run()

<class 'ragas.testset.synthesizers.testset_schema.Testset'>
21
                                           user_input  \
0   What role do fats play in a balanced diet acco...   
1   What role do proteins play in a balanced diet ...   
2   What role do minerals play in maintaining a ba...   
3                                     What are carbs?   
4   What role do carbohydrates play in a balanced ...   
5             What minerals are important for health?   
6   Can you tell me what PART 2 is about in that w...   
7   As a Holistic Wellness Coach, how can establis...   
8       What are some natural remedies for headaches?   
9   What are some natural ways to support digestiv...   
10                    How work-life balance help you?   
11  What are some natural ways to support digestiv...   
12  How can morning routines help improve your wel...   
13  As a Holistic Wellness Coach, how can understa...   
14  What are some recommended exercises for reliev...   
15  How does the shoulder

In [12]:
from typing import List, TypedDict, Dict, Callable, Any
import pickle
import time
from copy import deepcopy
import logging


from langchain.retrievers import ContextualCompressionRetriever, MultiQueryRetriever, ParentDocumentRetriever, \
    EnsembleRetriever
from langchain_community.document_loaders import TextLoader
from langchain_community.retrievers import BM25Retriever
from langchain_text_splitters import RecursiveCharacterTextSplitter
from ragas import RunConfig, evaluate, EvaluationDataset
from ragas.llms import LangchainLLMWrapper, BaseRagasLLM
from ragas.embeddings import LangchainEmbeddingsWrapper, BaseRagasEmbeddings
from langchain_openai import ChatOpenAI
from langchain_openai import OpenAIEmbeddings
from langchain_core.documents.base import Document
from ragas.testset import TestsetGenerator
from ragas.dataset_schema import EvaluationResult
from dotenv import load_dotenv
from ragas.testset.graph import KnowledgeGraph, Node, NodeType
from ragas.testset.transforms import default_transforms, apply_transforms
from ragas.testset.synthesizers import default_query_distribution, SingleHopSpecificQuerySynthesizer, MultiHopAbstractQuerySynthesizer, MultiHopSpecificQuerySynthesizer
from ragas.testset.synthesizers.testset_schema import Testset
from ragas.metrics import LLMContextRecall, Faithfulness, FactualCorrectness, ResponseRelevancy, ContextEntityRecall, \
    NoiseSensitivity
from langgraph.graph import START, StateGraph
from langchain_qdrant import QdrantVectorStore
from qdrant_client import QdrantClient, models
from qdrant_client.http.models import Distance, VectorParams
from langchain_core.retrievers import RetrieverLike, BaseRetriever
from langchain_cohere import CohereRerank
from langchain_core.language_models import BaseLanguageModel
from langchain.storage import InMemoryStore
from langchain_experimental.text_splitter import SemanticChunker
import pandas as pd


logger = logging.getLogger(__name__)

# load environment variables
load_dotenv()

HEALTH_AND_WELLNESS_GUIDE_PATH = "data/HealthWellnessGuide.txt"
GOLDEN_DATA_SET_FILE = "golden_data_set.pkl"
TEST_SET_SIZE = 15
CHUNK_SIZE = 500
CHUNK_OVERLAP = 50
PARENT_CHUNK_SIZE = 2000
PARENT_CHUNK_OVERLAP = 200
CHILD_CHUNK_SIZE = 400
CHILD_CHUNK_OVERLAP = 50
SLEEP_SECONDS = 1 # to avoid openAI rate-limiting
RAG_PROMPT = """
You are a helpful and kind assistant. Use the context provided below to answer the question.

If you do not know the answer, or are unsure, say you don't know.

### Question
{question}

### Context
{context}
"""

RAG_LLM = ChatOpenAI(model="gpt-4.1-mini")
RAG_EMBEDDING_MODEL = OpenAIEmbeddings(model="text-embedding-3-small")

def create_knowledge_graph(documents: List[Document], llm: BaseRagasLLM, embedding_model: BaseRagasEmbeddings,
                           cache_file: str = 'kg.json') -> KnowledgeGraph:
    """
    This function creates a Knowledge Graph (kg) of the documents supplied to it (this kg is utilized to create test queries downstream)
    Args:
        documents: the list of documents to create a knowledge graph of
        llm: the LLM to utilize to create transformations to our kg
        embedding_model: the embedding model to use with the LLM
        cache_file: file_path/name for the cached KG file (can reload this instead of redoing computation each time on same list of documents)

    Returns: KnowledgeGraph Object
    """
    # create knowledge graph
    try:
        kg = KnowledgeGraph.load(cache_file)
    except FileNotFoundError:
        kg = KnowledgeGraph()

        # add each document as a node in the KG
        for doc in documents:
            kg.nodes.append(
                Node(type=NodeType.DOCUMENT,
                     properties={"page_content": doc.page_content, "document_metadata": doc.metadata}))

        # apply transformations to KG (summarization, extracting headlines and themes..)
        transforms = default_transforms(documents=documents, llm=llm, embedding_model=embedding_model)
        apply_transforms(kg, transforms)
        kg.save(cache_file)
    return kg


def create_golden_data_set(documents: List[Document], llm: str, embedding_model: str,
                           use_cache: bool = True) -> Testset:
    # create knowledge graph from corpus (utilized for generating queries via LLM generator)
    # kg = create_knowledge_graph(documents, llm, embedding_model, cache_file='kg.json')
    try:
        test_queries = pickle.load(open(GOLDEN_DATA_SET_FILE, 'rb'))
    except FileNotFoundError:
        llm_wrapper = LangchainLLMWrapper(ChatOpenAI(model=llm))
        embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings(model=embedding_model))
        generator = TestsetGenerator(llm=llm_wrapper, embedding_model=embeddings)
        test_queries = generator.generate_with_langchain_docs(documents, testset_size=TEST_SET_SIZE)
        with open(GOLDEN_DATA_SET_FILE, 'wb') as f:
            pickle.dump(test_queries, f)


    # distribution of queries we want
    # single_specific, multi_abstract, multi_specific = query_type_ratios
    # query_distribution = [
    #     (SingleHopSpecificQuerySynthesizer(llm=llm), single_specific),
    #     (MultiHopAbstractQuerySynthesizer(llm=llm), multi_abstract),
    #     (MultiHopSpecificQuerySynthesizer(llm=llm), multi_specific),
    # ]

    return test_queries


class State(TypedDict):
    question: str
    context: List[Document]
    response: str

def generate(state: State):
    docs_content = "\n\n".join(doc.page_content for doc in state["context"])
    messages = RAG_PROMPT.format(question=state["question"], context=docs_content)
    response = RAG_LLM.invoke(messages)
    return {"response": response.content}

def get_naive_retriever(vector_store: QdrantVectorStore, k: int = 10, **kwargs):
    return vector_store.as_retriever(search_kwargs={"k": k})

def get_bm25_retriever(docs: List[Document], k: int = 10, **kwargs):
    return BM25Retriever.from_documents(docs, k=k)

def get_cohere_reranker(base_retriever: RetrieverLike, cohere_model: str = 'rerank-v3.5', **kwargs):
    compressor = CohereRerank(model=cohere_model)
    return ContextualCompressionRetriever(base_compressor=compressor, base_retriever=base_retriever)

def get_multiquery_retriever(retriever: BaseRetriever, llm: BaseLanguageModel, **kwargs):
    return MultiQueryRetriever.from_llm(retriever=retriever, llm=llm)

def get_parent_document_retriever(parent_chunk_size: int, parent_chunk_overlap: int, child_chunk_size: int,
                                  child_chunk_overlap: int, qdrant_client: QdrantClient, collection_name: str,
                                  documents: List[Document], openai_embedding_model: str = 'text-embedding-3-small',
                                  embedding_size: int = 1536, is_semantic: bool = False, **kwargs):
    qdrant_client.create_collection(
        collection_name=collection_name,
        vector_config=models.VectorConfig(size=embedding_size,
        distance=models.DistanceType.COSINE))
    parent_document_vector_store = QdrantVectorStore(
        collection_name=collection_name,
        embedding=OpenAIEmbeddings(model=openai_embedding_model),
        client=qdrant_client)

    parent_splitter = None if is_semantic else RecursiveCharacterTextSplitter(chunk_size=parent_chunk_size, overlap=parent_chunk_overlap)
    # create an InMemoryStore for the parent document retriever
    in_memory_store = InMemoryStore()
    parent_document_retriever = ParentDocumentRetriever(
        vectorstore=parent_document_vector_store,
        docstore=in_memory_store,
        child_splitter=RecursiveCharacterTextSplitter(chunk_size=child_chunk_size, overlap=child_chunk_overlap),
        parent_splitter=parent_splitter)
    parent_document_retriever.add_documents(documents)
    return parent_document_retriever

def load_sources(file_paths: List[str]):
    raw_documents = []

    # load all sources for RAG
    for file_path in file_paths:
        loader = TextLoader(file_path)
        raw_documents.extend(loader.load())
    return raw_documents

def get_ensemble_retriever(retrievers: List[RetrieverLike], weighting: List[float] = None, **kwargs):
    if weighting is None:
        weighting = [1.0 / len(retrievers)] * len(retrievers)
    else:
        assert (len(weighting) == len(retrievers)), "Length of weighting must be equal to length of retrievers!"
    return EnsembleRetriever(retrievers=retrievers, weights=weighting)

def set_up_vector_db(docs: List[Document], db_url: str = None, embedding_size: int = 1536, is_semantic: bool = False):
    # create vector store and add documents
    if db_url is None:
        db_url = ":memory:"

    if is_semantic:
        vector_store = QdrantVectorStore.from_documents(
            docs,
            RAG_EMBEDDING_MODEL,
            location=db_url,
            collection_name="wellness_semantic"
        )
        return None, vector_store
    else:
        client = QdrantClient(db_url)
        client.create_collection(
            collection_name="wellness",
            vectors_config=VectorParams(size=embedding_size, distance=Distance.COSINE)
        )
        vector_store = QdrantVectorStore(
            client=client,
            collection_name="wellness",
            embedding=RAG_EMBEDDING_MODEL
        )
        _ = vector_store.add_documents(docs)
        return client, vector_store

def retriever_factory(retriever: RetrieverLike):
    def retrieve_node(state: State):
        retrieved_docs = retriever.invoke(state["question"])
        return {"context": retrieved_docs}
    return retrieve_node

def build_rag_graph(retriever: Callable[[State], Dict[str, Any]]):
    rag_graph_builder = StateGraph(State).add_sequence([retriever, generate])
    rag_graph_builder.add_edge(START, "retrieve_node")  # node name from retriever_factory's __name__
    rag_graph = rag_graph_builder.compile()
    return rag_graph

def invoke_test_queries(rag_graph: StateGraph, dataset: Testset):
    # invoke each test query from the Testset object
    for test_row in dataset:
        response = rag_graph.invoke({"question": test_row.eval_sample.user_input})
        test_row.eval_sample.response = response["response"]
        test_row.eval_sample.retrieved_contexts = [context.page_content for context in response["context"]]
        time.sleep(SLEEP_SECONDS)  # to avoid rate limits
    return dataset

def gather_rag_statistics(data_set: Testset, llm: LangchainLLMWrapper):
    custom_run_config = RunConfig(timeout=60)
    results: EvaluationResult = evaluate(
        dataset=EvaluationDataset.from_pandas(data_set.to_pandas()),
        metrics=[LLMContextRecall(), Faithfulness(), FactualCorrectness(), ResponseRelevancy(), ContextEntityRecall(), NoiseSensitivity()],
        llm=llm,
        run_config=custom_run_config,
    )
    return results

def run_experiment_loop(golden_data_set: Testset, retriever_mapping: Dict[str, Callable], retriever_params: dict,
                        evaluator_llm: LangchainLLMWrapper, chunk_strategy: str) -> List[Dict]:
    rag_metrics = []
    for retriever_str, retriever_func in retriever_mapping.items():
        logger.warning(f"Beginning experiment loop: retriever: {retriever_str}, retriever_func: {retriever_func}, "
                       f"chunk_strategy: {chunk_strategy}")
        retriever_runnable: RetrieverLike = retriever_func(**retriever_params)

        # add retrievers (objects) to utilize for ensemble retriever at the end (mutating retriever_params dic)
        if retriever_str == "retriever":
            retriever_params.get("retrievers").append(retriever_runnable)

        retriever_node = retriever_factory(retriever=retriever_runnable)
        rag_graph = build_rag_graph(retriever=retriever_node)
        tested_data_set = invoke_test_queries(rag_graph, deepcopy(golden_data_set))
        rag_stats = gather_rag_statistics(tested_data_set, evaluator_llm)
        # EvaluationResult.scores is a list of per-row dicts; aggregate to mean per metric for this run
        scores_dict = pd.DataFrame(rag_stats.scores).mean(numeric_only=True, skipna=True).to_dict()
        rag_metrics.append({'retriever': retriever_str, 'chunking_strategy': f'{chunk_strategy}', **scores_dict})
        logger.warning(f"Done with experiment loop: retriever: {retriever_str}, retriever_func: {retriever_func.__name__}, "
                       f"chunk_strategy: {chunk_strategy}")
    return rag_metrics

def run_experiment(document_sources: List[str]):
    raw_corpus = load_sources(document_sources)

    # get golden data set for evaluation of retrieval methods
    golden_data_set: Testset = create_golden_data_set(raw_corpus, "gpt-4.1-nano", "text-embedding-3-small")

    text_splitter = RecursiveCharacterTextSplitter(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP)
    wellness_docs = text_splitter.split_documents(raw_corpus)
    vector_db_client, vector_store = set_up_vector_db(wellness_docs)

    naive_retriever = get_naive_retriever(vector_store=vector_store)
    non_semantic_params = {
        "vector_store": vector_store,                       # get_naive_retriever
        # "docs": wellness_docs,                              # get_bm25_retriever
        # "base_retriever": naive_retriever,                  # get_cohere_reranker
        "retriever": naive_retriever,                       # get_multiquery_retriever
        "llm": ChatOpenAI(model="gpt-4.1-nano"),            # get_multiquery_retriever
        # "parent_chunk_size": PARENT_CHUNK_SIZE,             # get_parent_document_retriever
        # "parent_chunk_overlap": PARENT_CHUNK_OVERLAP,       # get_parent_document_retriever
        # "child_chunk_size": CHILD_CHUNK_SIZE,               # get_parent_document_retriever
        # "child_chunk_overlap": CHILD_CHUNK_OVERLAP,         # get_parent_document_retriever
        # "qdrant_client": vector_db_client,                  # get_parent_document_retriever
        # "collection_name": "wellness_guide",                # get_parent_document_retriever
        # "documents": raw_corpus,                            # get_parent_document_retriever
        "retrievers": [],                                   # get_ensemble_retriever
    }

    retriever_mapping = {
        "naive": get_naive_retriever,
        # "bm25": get_bm25_retriever,
        # "reranker": get_cohere_reranker,
        "multiquery": get_multiquery_retriever,
        # "parent_document": get_parent_document_retriever,
        # "ensemble": get_ensemble_retriever,
    }

    evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4.1-mini"))

    non_semantic_rag_metrics = run_experiment_loop(
        golden_data_set, retriever_mapping, non_semantic_params, evaluator_llm, "non_semantic_chunking")

    # semantic chunking portion of experiment
    semantic_chunker = SemanticChunker(RAG_EMBEDDING_MODEL)
    semantic_wellness_docs = semantic_chunker.split_documents(wellness_docs)
    _, semantic_vector_store = set_up_vector_db(semantic_wellness_docs, is_semantic=True)

    naive_semantic_retriever = get_naive_retriever(semantic_vector_store)
    semantic_params = {
        "vector_store": semantic_vector_store,              # get_naive_retriever
        # "docs": semantic_wellness_docs,                     # get_bm25_retriever
        # "base_retriever": naive_semantic_retriever,         # get_cohere_reranker
        # "retriever": naive_semantic_retriever,              # get_multiquery_retriever
        # "llm": ChatOpenAI(model="gpt-4.1-nano"),            # get_multiquery_retriever
        # "parent_chunk_size": PARENT_CHUNK_SIZE,             # get_parent_document_retriever
        # "parent_chunk_overlap": PARENT_CHUNK_OVERLAP,       # get_parent_document_retriever
        # "child_chunk_size": CHILD_CHUNK_SIZE,               # get_parent_document_retriever
        # "child_chunk_overlap": CHILD_CHUNK_OVERLAP,         # get_parent_document_retriever
        # "qdrant_client": vector_db_client,                  # get_parent_document_retriever
        # "collection_name": "semantic_wellness_guide",       # get_parent_document_retriever
        # "documents": semantic_wellness_docs,                # get_parent_document_retriever
        "retrievers": [],                                   # get_ensemble_retriever
    }

    semantic_mapping = {
        "naive": get_naive_retriever,
        # "bm25": get_bm25_retriever,
        # "reranker": get_cohere_reranker,
        # "multiquery": get_multiquery_retriever,
        # "parent_document": get_parent_document_retriever,
        # "ensemble": get_ensemble_retriever,
    }

    semantic_rag_metrics = run_experiment_loop(
        golden_data_set, semantic_mapping, semantic_params, evaluator_llm, "semantic_chunking")

    # turn metrics into a dataframe for easier analysis and plotting
    metrics_df = pd.DataFrame(non_semantic_rag_metrics + semantic_rag_metrics)
    return metrics_df



if __name__ == "__main__":
    metrics_df = run_experiment(document_sources=[f"{HEALTH_AND_WELLNESS_GUIDE_PATH}"])

In [13]:
metrics_df = run_experiment(document_sources=[f"{HEALTH_AND_WELLNESS_GUIDE_PATH}"])

Applying HeadlinesExtractor:   0%|          | 0/1 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/1 [00:00<?, ?it/s]

Applying SummaryExtractor:   0%|          | 0/1 [00:00<?, ?it/s]

Applying CustomNodeFilter:   0%|          | 0/3 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/7 [00:00<?, ?it/s]

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

Generating personas:   0%|          | 0/1 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/2 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/6 [00:00<?, ?it/s]

Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x16889a5c0>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

Exception raised in Job[11]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x16889a5c0>, chunk_strategy: semantic_chunking


Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

Exception raised in Job[11]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: semantic_chunking


AttributeError: module 'pandas' has no attribute 'Dataframe'

In [20]:
from typing import List, TypedDict, Dict, Callable, Any
import pickle
import time
from copy import deepcopy
import logging


from langchain.retrievers import ContextualCompressionRetriever, MultiQueryRetriever, ParentDocumentRetriever, \
    EnsembleRetriever
from langchain_community.document_loaders import TextLoader
from langchain_community.retrievers import BM25Retriever
from langchain_text_splitters import RecursiveCharacterTextSplitter
from ragas import RunConfig, evaluate, EvaluationDataset
from ragas.llms import LangchainLLMWrapper, BaseRagasLLM
from ragas.embeddings import LangchainEmbeddingsWrapper, BaseRagasEmbeddings
from langchain_openai import ChatOpenAI
from langchain_openai import OpenAIEmbeddings
from langchain_core.documents.base import Document
from ragas.testset import TestsetGenerator
from ragas.dataset_schema import EvaluationResult
from dotenv import load_dotenv
from ragas.testset.graph import KnowledgeGraph, Node, NodeType
from ragas.testset.transforms import default_transforms, apply_transforms
from ragas.testset.synthesizers import default_query_distribution, SingleHopSpecificQuerySynthesizer, MultiHopAbstractQuerySynthesizer, MultiHopSpecificQuerySynthesizer
from ragas.testset.synthesizers.testset_schema import Testset
from ragas.metrics import LLMContextRecall, Faithfulness, FactualCorrectness, ResponseRelevancy, ContextEntityRecall, \
    NoiseSensitivity
from langgraph.graph import START, StateGraph
from langchain_qdrant import QdrantVectorStore
from qdrant_client import QdrantClient, models
from qdrant_client.http.models import Distance, VectorParams
from langchain_core.retrievers import RetrieverLike, BaseRetriever
from langchain_cohere import CohereRerank
from langchain_core.language_models import BaseLanguageModel
from langchain.storage import InMemoryStore
from langchain_experimental.text_splitter import SemanticChunker
import pandas as pd


logger = logging.getLogger(__name__)

# load environment variables
load_dotenv()

HEALTH_AND_WELLNESS_GUIDE_PATH = "data/HealthWellnessGuide.txt"
GOLDEN_DATA_SET_FILE = "golden_data_set.pkl"
TEST_SET_SIZE = 15
CHUNK_SIZE = 500
CHUNK_OVERLAP = 50
PARENT_CHUNK_SIZE = 2000
PARENT_CHUNK_OVERLAP = 200
CHILD_CHUNK_SIZE = 400
CHILD_CHUNK_OVERLAP = 50
SLEEP_SECONDS = 1 # to avoid openAI rate-limiting
RAG_PROMPT = """
You are a helpful and kind assistant. Use the context provided below to answer the question.

If you do not know the answer, or are unsure, say you don't know.

### Question
{question}

### Context
{context}
"""

RAG_LLM = ChatOpenAI(model="gpt-4.1-mini")
RAG_EMBEDDING_MODEL = OpenAIEmbeddings(model="text-embedding-3-small")

def create_knowledge_graph(documents: List[Document], llm: BaseRagasLLM, embedding_model: BaseRagasEmbeddings,
                           cache_file: str = 'kg.json') -> KnowledgeGraph:
    """
    This function creates a Knowledge Graph (kg) of the documents supplied to it (this kg is utilized to create test queries downstream)
    Args:
        documents: the list of documents to create a knowledge graph of
        llm: the LLM to utilize to create transformations to our kg
        embedding_model: the embedding model to use with the LLM
        cache_file: file_path/name for the cached KG file (can reload this instead of redoing computation each time on same list of documents)

    Returns: KnowledgeGraph Object
    """
    # create knowledge graph
    try:
        kg = KnowledgeGraph.load(cache_file)
    except FileNotFoundError:
        kg = KnowledgeGraph()

        # add each document as a node in the KG
        for doc in documents:
            kg.nodes.append(
                Node(type=NodeType.DOCUMENT,
                     properties={"page_content": doc.page_content, "document_metadata": doc.metadata}))

        # apply transformations to KG (summarization, extracting headlines and themes..)
        transforms = default_transforms(documents=documents, llm=llm, embedding_model=embedding_model)
        apply_transforms(kg, transforms)
        kg.save(cache_file)
    return kg


def create_golden_data_set(documents: List[Document], llm: str, embedding_model: str,
                           use_cache: bool = True) -> Testset:
    # create knowledge graph from corpus (utilized for generating queries via LLM generator)
    # kg = create_knowledge_graph(documents, llm, embedding_model, cache_file='kg.json')
    try:
        test_queries = pickle.load(open(GOLDEN_DATA_SET_FILE, 'rb'))
    except FileNotFoundError:
        llm_wrapper = LangchainLLMWrapper(ChatOpenAI(model=llm))
        embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings(model=embedding_model))
        generator = TestsetGenerator(llm=llm_wrapper, embedding_model=embeddings)
        test_queries = generator.generate_with_langchain_docs(documents, testset_size=TEST_SET_SIZE)
        with open(GOLDEN_DATA_SET_FILE, 'wb') as f:
            pickle.dump(test_queries, f)


    # distribution of queries we want
    # single_specific, multi_abstract, multi_specific = query_type_ratios
    # query_distribution = [
    #     (SingleHopSpecificQuerySynthesizer(llm=llm), single_specific),
    #     (MultiHopAbstractQuerySynthesizer(llm=llm), multi_abstract),
    #     (MultiHopSpecificQuerySynthesizer(llm=llm), multi_specific),
    # ]

    return test_queries


class State(TypedDict):
    question: str
    context: List[Document]
    response: str

def generate(state: State):
    docs_content = "\n\n".join(doc.page_content for doc in state["context"])
    messages = RAG_PROMPT.format(question=state["question"], context=docs_content)
    response = RAG_LLM.invoke(messages)
    return {"response": response.content}

def get_naive_retriever(vector_store: QdrantVectorStore, k: int = 10, **kwargs):
    return vector_store.as_retriever(search_kwargs={"k": k})

def get_bm25_retriever(docs: List[Document], k: int = 10, **kwargs):
    return BM25Retriever.from_documents(docs, k=k)

def get_cohere_reranker(base_retriever: RetrieverLike, cohere_model: str = 'rerank-v3.5', **kwargs):
    compressor = CohereRerank(model=cohere_model)
    return ContextualCompressionRetriever(base_compressor=compressor, base_retriever=base_retriever)

def get_multiquery_retriever(retriever: BaseRetriever, llm: BaseLanguageModel, **kwargs):
    return MultiQueryRetriever.from_llm(retriever=retriever, llm=llm)

def get_parent_document_retriever(parent_chunk_size: int, parent_chunk_overlap: int, child_chunk_size: int,
                                  child_chunk_overlap: int, qdrant_client: QdrantClient, collection_name: str,
                                  documents: List[Document], openai_embedding_model: str = 'text-embedding-3-small',
                                  embedding_size: int = 1536, is_semantic: bool = False, **kwargs):
    qdrant_client.create_collection(
        collection_name=collection_name,
        vector_config=models.VectorConfig(size=embedding_size,
        distance=models.DistanceType.COSINE))
    parent_document_vector_store = QdrantVectorStore(
        collection_name=collection_name,
        embedding=OpenAIEmbeddings(model=openai_embedding_model),
        client=qdrant_client)

    parent_splitter = None if is_semantic else RecursiveCharacterTextSplitter(chunk_size=parent_chunk_size, overlap=parent_chunk_overlap)
    # create an InMemoryStore for the parent document retriever
    in_memory_store = InMemoryStore()
    parent_document_retriever = ParentDocumentRetriever(
        vectorstore=parent_document_vector_store,
        docstore=in_memory_store,
        child_splitter=RecursiveCharacterTextSplitter(chunk_size=child_chunk_size, overlap=child_chunk_overlap),
        parent_splitter=parent_splitter)
    parent_document_retriever.add_documents(documents)
    return parent_document_retriever

def load_sources(file_paths: List[str]):
    raw_documents = []

    # load all sources for RAG
    for file_path in file_paths:
        loader = TextLoader(file_path)
        raw_documents.extend(loader.load())
    return raw_documents

def get_ensemble_retriever(retrievers: List[RetrieverLike], weighting: List[float] = None, **kwargs):
    if weighting is None:
        weighting = [1.0 / len(retrievers)] * len(retrievers)
    else:
        assert (len(weighting) == len(retrievers)), "Length of weighting must be equal to length of retrievers!"
    return EnsembleRetriever(retrievers=retrievers, weights=weighting)

def set_up_vector_db(docs: List[Document], db_url: str = None, embedding_size: int = 1536, is_semantic: bool = False):
    # create vector store and add documents
    if db_url is None:
        db_url = ":memory:"

    if is_semantic:
        vector_store = QdrantVectorStore.from_documents(
            docs,
            RAG_EMBEDDING_MODEL,
            location=db_url,
            collection_name="wellness_semantic"
        )
        return None, vector_store
    else:
        client = QdrantClient(db_url)
        client.create_collection(
            collection_name="wellness",
            vectors_config=VectorParams(size=embedding_size, distance=Distance.COSINE)
        )
        vector_store = QdrantVectorStore(
            client=client,
            collection_name="wellness",
            embedding=RAG_EMBEDDING_MODEL
        )
        _ = vector_store.add_documents(docs)
        return client, vector_store

def retriever_factory(retriever: RetrieverLike):
    def retrieve_node(state: State):
        retrieved_docs = retriever.invoke(state["question"])
        return {"context": retrieved_docs}
    return retrieve_node

def build_rag_graph(retriever: Callable[[State], Dict[str, Any]]):
    rag_graph_builder = StateGraph(State).add_sequence([retriever, generate])
    rag_graph_builder.add_edge(START, "retrieve_node")  # node name from retriever_factory's __name__
    rag_graph = rag_graph_builder.compile()
    return rag_graph

def invoke_test_queries(rag_graph: StateGraph, dataset: Testset):
    # invoke each test query from the Testset object
    for test_row in dataset:
        response = rag_graph.invoke({"question": test_row.eval_sample.user_input})
        test_row.eval_sample.response = response["response"]
        test_row.eval_sample.retrieved_contexts = [context.page_content for context in response["context"]]
        time.sleep(SLEEP_SECONDS)  # to avoid rate limits
    return dataset

def gather_rag_statistics(data_set: Testset, llm: LangchainLLMWrapper):
    custom_run_config = RunConfig(timeout=60)
    results: EvaluationResult = evaluate(
        dataset=EvaluationDataset.from_pandas(data_set.to_pandas()),
        metrics=[LLMContextRecall(), Faithfulness(), FactualCorrectness(), ResponseRelevancy(), ContextEntityRecall(), NoiseSensitivity()],
        llm=llm,
        run_config=custom_run_config,
    )
    return results

def run_experiment_loop(golden_data_set: Testset, retriever_mapping: Dict[str, Callable], retriever_params: dict,
                        evaluator_llm: LangchainLLMWrapper, chunk_strategy: str) -> List[Dict]:
    rag_metrics = []
    for retriever_str, retriever_func in retriever_mapping.items():
        logger.warning(f"Beginning experiment loop: retriever: {retriever_str}, retriever_func: {retriever_func}, "
                       f"chunk_strategy: {chunk_strategy}")
        retriever_runnable: RetrieverLike = retriever_func(**retriever_params)

        # add retrievers (objects) to utilize for ensemble retriever at the end (mutating retriever_params dic)
        if retriever_str == "retriever":
            retriever_params.get("retrievers").append(retriever_runnable)

        retriever_node = retriever_factory(retriever=retriever_runnable)
        rag_graph = build_rag_graph(retriever=retriever_node)
        tested_data_set = invoke_test_queries(rag_graph, deepcopy(golden_data_set))
        rag_stats = gather_rag_statistics(tested_data_set, evaluator_llm)
        # EvaluationResult.scores is a list of per-row dicts; aggregate to mean per metric for this run
        scores_dict = pd.DataFrame(rag_stats.scores).mean(numeric_only=True, skipna=True).to_dict()
        rag_metrics.append({'retriever': retriever_str, 'chunking_strategy': f'{chunk_strategy}', **scores_dict})
        logger.warning(f"Done with experiment loop: retriever: {retriever_str}, retriever_func: {retriever_func.__name__}, "
                       f"chunk_strategy: {chunk_strategy}")
    return rag_metrics

def run_experiment(document_sources: List[str]):
    raw_corpus = load_sources(document_sources)

    # get golden data set for evaluation of retrieval methods
    golden_data_set: Testset = create_golden_data_set(raw_corpus, "gpt-4.1-nano", "text-embedding-3-small")

    text_splitter = RecursiveCharacterTextSplitter(chunk_size=CHUNK_SIZE, chunk_overlap=CHUNK_OVERLAP)
    wellness_docs = text_splitter.split_documents(raw_corpus)
    vector_db_client, vector_store = set_up_vector_db(wellness_docs)

    naive_retriever = get_naive_retriever(vector_store=vector_store)
    non_semantic_params = {
        "vector_store": vector_store,                       # get_naive_retriever
        # "docs": wellness_docs,                              # get_bm25_retriever
        # "base_retriever": naive_retriever,                  # get_cohere_reranker
        "retriever": naive_retriever,                       # get_multiquery_retriever
        "llm": ChatOpenAI(model="gpt-4.1-nano"),            # get_multiquery_retriever
        # "parent_chunk_size": PARENT_CHUNK_SIZE,             # get_parent_document_retriever
        # "parent_chunk_overlap": PARENT_CHUNK_OVERLAP,       # get_parent_document_retriever
        # "child_chunk_size": CHILD_CHUNK_SIZE,               # get_parent_document_retriever
        # "child_chunk_overlap": CHILD_CHUNK_OVERLAP,         # get_parent_document_retriever
        # "qdrant_client": vector_db_client,                  # get_parent_document_retriever
        # "collection_name": "wellness_guide",                # get_parent_document_retriever
        # "documents": raw_corpus,                            # get_parent_document_retriever
        "retrievers": [],                                   # get_ensemble_retriever
    }

    retriever_mapping = {
        "naive": get_naive_retriever,
        # "bm25": get_bm25_retriever,
        # "reranker": get_cohere_reranker,
        "multiquery": get_multiquery_retriever,
        # "parent_document": get_parent_document_retriever,
        # "ensemble": get_ensemble_retriever,
    }

    evaluator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4.1-mini"))

    non_semantic_rag_metrics = run_experiment_loop(
        golden_data_set, retriever_mapping, non_semantic_params, evaluator_llm, "non_semantic_chunking")

    # semantic chunking portion of experiment
    semantic_chunker = SemanticChunker(RAG_EMBEDDING_MODEL)
    semantic_wellness_docs = semantic_chunker.split_documents(wellness_docs)
    _, semantic_vector_store = set_up_vector_db(semantic_wellness_docs, is_semantic=True)

    naive_semantic_retriever = get_naive_retriever(semantic_vector_store)
    semantic_params = {
        "vector_store": semantic_vector_store,              # get_naive_retriever
        # "docs": semantic_wellness_docs,                     # get_bm25_retriever
        # "base_retriever": naive_semantic_retriever,         # get_cohere_reranker
        # "retriever": naive_semantic_retriever,              # get_multiquery_retriever
        # "llm": ChatOpenAI(model="gpt-4.1-nano"),            # get_multiquery_retriever
        # "parent_chunk_size": PARENT_CHUNK_SIZE,             # get_parent_document_retriever
        # "parent_chunk_overlap": PARENT_CHUNK_OVERLAP,       # get_parent_document_retriever
        # "child_chunk_size": CHILD_CHUNK_SIZE,               # get_parent_document_retriever
        # "child_chunk_overlap": CHILD_CHUNK_OVERLAP,         # get_parent_document_retriever
        # "qdrant_client": vector_db_client,                  # get_parent_document_retriever
        # "collection_name": "semantic_wellness_guide",       # get_parent_document_retriever
        # "documents": semantic_wellness_docs,                # get_parent_document_retriever
        "retrievers": [],                                   # get_ensemble_retriever
    }

    semantic_mapping = {
        "naive": get_naive_retriever,
        # "bm25": get_bm25_retriever,
        # "reranker": get_cohere_reranker,
        # "multiquery": get_multiquery_retriever,
        # "parent_document": get_parent_document_retriever,
        # "ensemble": get_ensemble_retriever,
    }

    semantic_rag_metrics = run_experiment_loop(
        golden_data_set, semantic_mapping, semantic_params, evaluator_llm, "semantic_chunking")

    # turn metrics into a dataframe for easier analysis and plotting
    metrics_df = pd.DataFrame(non_semantic_rag_metrics + semantic_rag_metrics)
    return metrics_df

In [16]:
metrics_df = run_experiment(document_sources=[f"{HEALTH_AND_WELLNESS_GUIDE_PATH}"])

Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x14c82cb80>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[19]: TimeoutError()
Exception raised in Job[20]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[31]: TimeoutError()
Exception raised in Job[32]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: multiquery, retriever_func: <function get_multiquery_retriever at 0x16889afc0>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[25]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[32]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Done with experiment loop: retriever: multiquery, retriever_func: get_multiquery_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x14c82cb80>, chunk_strategy: semantic_chunking


Evaluating:   0%|          | 0/36 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[19]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[32]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: semantic_chunking


In [17]:
# this run was a trial run with just a test set size of 5
metrics_df

retriever      chunking_strategy  context_recall  faithfulness  \
0       naive  non_semantic_chunking        0.888889      0.821918   
1  multiquery  non_semantic_chunking        0.944444      0.738017   
2       naive      semantic_chunking        0.777778      0.758282   

   factual_correctness  answer_relevancy  context_entity_recall  \
0               0.4975          0.962616               0.369298   
1               0.5000          0.971287               0.341082   
2               0.4660          0.973520               0.337427   

   noise_sensitivity_relevant  
0                         NaN  
1                         NaN  
2                         NaN

- this was a test run with just a test set of 5 (so I feel the numbers are a bit noisy)
- let's try with a test set size of **15** which is more statistically sound

In [18]:
metrics_df2 = run_experiment(document_sources=[f"{HEALTH_AND_WELLNESS_GUIDE_PATH}"])

Applying HeadlinesExtractor:   0%|          | 0/1 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/1 [00:00<?, ?it/s]

Applying SummaryExtractor:   0%|          | 0/1 [00:00<?, ?it/s]

Applying CustomNodeFilter:   0%|          | 0/3 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/7 [00:00<?, ?it/s]

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

Generating personas:   0%|          | 0/1 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/1 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/15 [00:00<?, ?it/s]

Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x14c82cb80>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[8]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Exception raised in Job[41]: TimeoutError()
Exception raised in Job[47]: TimeoutError()
Exception raised in Job[53]: TimeoutError()
Exception raised in Job[59]: TimeoutError()
Exception raised in Job[65]: TimeoutError()
Exception raised in Job[71]: TimeoutError()
Exception raised in Job[77]: TimeoutError()
Exception raised in Job[83]: TimeoutError()
Exception raised in Job[89]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: multiquery, retriever_func: <function get_multiquery_retriever at 0x16889afc0>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[7]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Exception raised in Job[41]: TimeoutError()
Exception raised in Job[47]: TimeoutError()
Exception raised in Job[53]: TimeoutError()
Exception raised in Job[59]: TimeoutError()
Exception raised in Job[65]: TimeoutError()
Exception raised in Job[71]: TimeoutError()
Exception raised in Job[73]: TimeoutError()
Exception raised in Job[77]: TimeoutError()
Exception raised in Job[83]: TimeoutError()
Exception raised in Job[89]: TimeoutError()
Done with experiment loop: retriever: multiquery, retriever_func: get_multiquery_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x14c82cb80>, chunk_strategy: 

Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Exception raised in Job[41]: TimeoutError()
Exception raised in Job[47]: TimeoutError()
Exception raised in Job[53]: TimeoutError()
Exception raised in Job[59]: TimeoutError()
Exception raised in Job[65]: TimeoutError()
Exception raised in Job[71]: TimeoutError()
Exception raised in Job[73]: TimeoutError()
Exception raised in Job[74]: TimeoutError()
Exception raised in Job[77]: TimeoutError()
Exception raised in Job[83]: TimeoutError()
Exception raised in Job[89]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: semantic_chunking


In [19]:
metrics_df2

retriever      chunking_strategy  context_recall  faithfulness  \
0       naive  non_semantic_chunking        0.877778      0.850064   
1  multiquery  non_semantic_chunking        0.922222      0.886434   
2       naive      semantic_chunking        0.844444      0.859176   

   factual_correctness  answer_relevancy  context_entity_recall  \
0             0.520714          0.836143               0.288691   
1             0.544000          0.901995               0.306325   
2             0.620714          0.974841               0.324565   

   noise_sensitivity_relevant  
0                         NaN  
1                         NaN  
2                         NaN

In [21]:
metrics_df3 = run_experiment(document_sources=[f"{HEALTH_AND_WELLNESS_GUIDE_PATH}"])

Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_retriever at 0x16ace00e0>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Exception raised in Job[41]: TimeoutError()
Exception raised in Job[47]: TimeoutError()
Exception raised in Job[53]: TimeoutError()
Exception raised in Job[59]: TimeoutError()
Exception raised in Job[65]: TimeoutError()
Exception raised in Job[71]: TimeoutError()
Exception raised in Job[77]: TimeoutError()
Exception raised in Job[83]: TimeoutError()
Exception raised in Job[89]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: multiquery, retriever_func: <function get_multiquery_retriever at 0x16ace0680>, chunk_strategy: non_semantic_chunking


Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[7]: TimeoutError()
Exception raised in Job[8]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Exception raised in Job[40]: TimeoutError()
Exception raised in Job[41]: TimeoutError()
Exception raised in Job[47]: TimeoutError()
Exception raised in Job[53]: TimeoutError()
Exception raised in Job[59]: TimeoutError()
Exception raised in Job[65]: TimeoutError()
Exception raised in Job[71]: TimeoutError()
Exception raised in Job[77]: TimeoutError()
Exception raised in Job[83]: TimeoutError()
Exception raised in Job[89]: TimeoutError()
Done with experiment loop: retriever: multiquery, retriever_func: get_multiquery_retriever, chunk_strategy: non_semantic_chunking
Beginning experiment loop: retriever: naive, retriever_func: <function get_naive_

Evaluating:   0%|          | 0/90 [00:00<?, ?it/s]

Exception raised in Job[5]: TimeoutError()
Exception raised in Job[11]: TimeoutError()
Exception raised in Job[17]: TimeoutError()
Exception raised in Job[23]: TimeoutError()
Exception raised in Job[29]: TimeoutError()
Exception raised in Job[35]: TimeoutError()
Exception raised in Job[41]: TimeoutError()
Exception raised in Job[47]: TimeoutError()
Exception raised in Job[53]: TimeoutError()
Exception raised in Job[59]: TimeoutError()
Exception raised in Job[65]: TimeoutError()
Exception raised in Job[71]: TimeoutError()
Exception raised in Job[73]: TimeoutError()
Exception raised in Job[77]: TimeoutError()
Exception raised in Job[83]: TimeoutError()
Exception raised in Job[89]: TimeoutError()
Done with experiment loop: retriever: naive, retriever_func: get_naive_retriever, chunk_strategy: semantic_chunking


In [22]:
metrics_df3

retriever      chunking_strategy  context_recall  faithfulness  \
0       naive  non_semantic_chunking        0.877778      0.867838   
1  multiquery  non_semantic_chunking        0.900000      0.839957   
2       naive      semantic_chunking        0.827778      0.880908   

   factual_correctness  answer_relevancy  context_entity_recall  \
0             0.555333          0.833617               0.288030   
1             0.502857          0.902770               0.297127   
2             0.533333          0.960202               0.306648   

   noise_sensitivity_relevant  
0                         NaN  
1                         NaN  
2                         NaN

 **Multi-Query Retriever Performance**
  - Multi-query achieved the highest context recall (0.90), showing it successfully retrieved more relevant information by generating multiple query variations
  - However, it had the lowest factual correctness (0.503), suggesting that casting a wider net retrieved more context but potentially introduced less precise information
  - Strong answer relevancy (0.903) indicates the retrieved context was still relevant to answering the questions
  - multi-query retriever definitely requires more performance considerations (since you have to generate more query versions from an LLM)
    - so testing this method comes at a cost of price & latency (price due to more LLM API calls and latency due to need time to generate the other query formulations)

 **Semantic Chunking**
  - Semantic chunking with naive retrieval showed the highest answer relevancy (0.960) and best faithfulness (0.881), meaning answers were more relevant and grounded in the retrieved context
  - However, it had the lowest context recall (0.828), suggesting semantic boundaries created chunks that were missed during retrieval
  - Best context entity recall (0.307) shows semantic chunking better preserved entity relationships within chunks

**Trade-offs**
  - Recall vs. Precision trade-off: Multi-query gets more context (high recall) but lower factual correctness, while naive retrieval is more precise but retrieves less
  - Chunking strategy matters: Semantic chunking improved answer quality metrics (relevancy, faithfulness) but at the cost of recall
  - no method by itself was an obvious winner; each approach was better for different metrics which reinforces the idea that you have to test different retrieval strategies for your own RAG application and its unique data sources

**Takeaways**
  - If your use case requires comprehensive information retrieval, multi-query is superior
  - If answer accuracy and relevance are critical, semantic chunking with proper retrieval tuning may be better
  - The "best" retriever depends on whether you optimize for recall, precision, or answer quality
  - it would be interesting to test semantic chunking with multiquery retrieval to see how the retrieval metrics would have compared (that would be my next experiment)